session/pingpong/invoice_tracker.go

```diff
@@ -143,18 +242,74 @@ func (it *InvoiceTracker) handleExchangeMessageReceiveError(err error) error {
 	return err
 }
 
-func (it *InvoiceTracker) sendInvoice() error {
-	// TODO: a ton of actions should go here
+func (it *InvoiceTracker) incrementAccountantFailureCount() {
+	atomic.AddUint64(&it.accountantFailureCount, 1)
+}
+
+func (it *InvoiceTracker) resetAccountantFailureCount() {
+	atomic.SwapUint64(&it.accountantFailureCount, 0)
+}
+
+func (it *InvoiceTracker) getAccountantFailureCount() uint64 {
+	return atomic.LoadUint64(&it.accountantFailureCount)
+}
+
+func (it *InvoiceTracker) validateExchangeMessage(em crypto.ExchangeMessage) error {
+	if res := em.ValidateExchangeMessage(common.HexToAddress(it.peer.Address)); !res {
+		return ErrExchangeValidationFailed
+	}
+
+	if em.Promise.Amount < it.lastExchangeMessage.Promise.Amount {
+		log.Warnf("consumer sent an invalid amount. Expected < %v, got %v", it.lastExchangeMessage.Promise.Amount, em.Promise.Amount)
+		return errors.Wrap(ErrConsumerPromiseValidationFailed, "invalid amount")
+	}
+
+	if em.Promise.Hashlock != it.lastInvoice.invoice.Hashlock {
+		log.Warnf("consumer sent an invalid hashlock. Expected %q, got %q", it.lastInvoice.invoice.Hashlock, em.Promise.Hashlock)
+		return errors.Wrap(ErrConsumerPromiseValidationFailed, "missmatching hashlock")
+	}
+
+	addr, err := crypto.GenerateChannelAddress(it.peer.Address, it.registryAddress, it.channelImplementation)
+	if err != nil {
+		return errors.Wrap(err, "could not generate channel address")
+	}
 
-	// TODO: fill the fields
-	return it.peerInvoiceSender.Send(crypto.Invoice{AgreementID: 1234})
+	if strings.ToLower(em.Promise.ChannelID) != strings.ToLower(addr) {
+		log.Warnf("consumer sent an invalid channel address. Expected %q, got %q", addr, em.Promise.ChannelID)
+		return errors.Wrap(ErrConsumerPromiseValidationFailed, "invalid channel address")
+	}
+	return nil
 }
 
 func (it *InvoiceTracker) receiveExchangeMessageOrTimeout() error {
 	select {
 	case pm := <-it.exchangeMessageChan:
-		if res := pm.ValidateExchangeMessage(common.HexToAddress(it.peer.Address)); !res {
-			return ErrExchangeValidationFailed
+		err := it.validateExchangeMessage(pm)
+		if err != nil {
+			return err
+		}
+
+		it.lastExchangeMessage = pm
+
+		promise, err := it.accountantCaller.RequestPromise(pm)
+		if err != nil {
+			it.incrementAccountantFailureCount()
+			if it.getAccountantFailureCount() > 3 {
+				return errors.Wrap(err, "could not call accountant")
+			}
+			return nil
+		}
+		it.resetAccountantFailureCount()
+		err = it.accountantPromiseStorage.Store(it.accountantID, promise)
+		if err != nil {
+			return errors.Wrap(err, "could not store accountant promise")
+		}
+		log.Debug("accountant promise stored")
+		hexR := hex.EncodeToString(it.lastInvoice.r)
+		err = it.accountantCaller.RevealR(hexR, it.providerID.Address, it.lastInvoice.invoice.AgreementID)
+		if err != nil {
+			// TODO: need to think about handling this a bit better
```

Comment:
Provider should check failure message and act accordingly. Possible scenarios:

 - 500 of any kind or server is not responding --> retry in one minute
 - 400 "R have to be provided as proper hex" --> ??? don't know how it could happen  here
 - 400 of any other kind --> try to `request_promise` (with same exchange message) once again and try revealing R again.

P.S. should we have all that messages somewhere in one package as constant and reuse in node and in accountant?

Reply:
#1346